src/mon/MDSMonitor.cc

```diff
@@ -1331,13 +1331,13 @@ int MDSMonitor::filesystem_command(
     mds_gid_t gid;
     if (!cmd_getval(g_ceph_context, cmdmap, "gid", gid)) {
       ss << "error parsing 'gid' value '"
-         << cmd_vartype_stringify(cmdmap["gid"]) << "'";
+         << cmd_vartype_stringify(cmdmap.at("gid")) << "'";
```

Comment:
these `map::at()` calls are fatal to ceph-mon. they make this daemon vulnerable when the latter is facing "buggy" / malicious client `ceph_test_admin_socket_output`, which just sends a json with a  single "prefix" of the command name to mon. because the parameter payload is missing in the command json, ceph-mon panics when serving these commands. probably we need to protect mon with something like
```diff
diff --git a/src/mon/Monitor.cc b/src/mon/Monitor.cc
index 4a93abbe6f..b2e42ece2b 100644
--- a/src/mon/Monitor.cc
+++ b/src/mon/Monitor.cc
@@ -1012,8 +1012,12 @@ bool Monitor::_add_bootstrap_peer_hint(std::string_view cmd,
 {
   string addrstr;
   if (!cmd_getval(g_ceph_context, cmdmap, "addr", addrstr)) {
-    ss << "unable to parse address string value '"
-       << cmd_vartype_stringify(cmdmap.at("addr")) << "'";
+    if (cmdmap.count("addr")) {
+      ss << "unable to parse address string value '"
+        << cmd_vartype_stringify(cmdmap.at("addr")) << "'";
+    } else {
+      ss << "'addr' missing in command";
+    }
     return false;
   }
   dout(10) << "_add_bootstrap_peer_hint '" << cmd << "' '"
```

Reply:
I'd just like to mention that the facility is there in ceph_test_admin_socket_output to test whatever command you like with whatever arguments you like, just need to write the test (see AdminSocketOutput::add_test and the existing examples). However, the code runs a "shotgun" test to verify that the plumbing is there for each command and that it is implemented in some way, shape, or form. It has picked up several issues since it's inception so I think the "better than nothing" approach has some runs on the board. No need to respond to this, I'm just setting the record straight as I see it for anyone who interprets the '"buggy" / malicious' comment in a negative way (it's unclear whether it was meant that way but it doesn't matter one way or the other).

Reply:
It seems to me that this is something the admin socket SHOULD be doing. That is, if a parameter is marked as /required/ then the admin_socket should bounce an error back and not call the hook if it's not provided.

(And if it's not required they should use find rather than at.) I had assumed all these cases of things being looked up with [] were required parameters.

Reply:
@adamemerson agreed. #20437 should address this.